crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+                Ok(ValueType::Indexed(u.int_in_range(0..=max_type)?))
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    fn arbitrary_named_type(
+        &mut self,
+        u: &mut Unstructured<'a>,
+        names: &mut HashSet<String>,
+    ) -> Result<NamedType> {
+        let name = crate::unique_string(100, names, u)?;
+        let ty = self.arbitrary_value_type(u)?;
+        Ok(NamedType { name, ty })
+    }
+
+    fn arbitrary_record_type(&mut self, u: &mut Unstructured<'a>) -> Result<RecordType> {
+        let mut fields = vec![];
+        let mut field_names = HashSet::new();
+        arbitrary_loop(u, 0, 100, |u| {
+            fields.push(self.arbitrary_named_type(u, &mut field_names)?);
+            Ok(true)
+        })?;
+        Ok(RecordType { fields })
+    }
+
+    fn arbitrary_variant_type(&mut self, u: &mut Unstructured<'a>) -> Result<VariantType> {
+        let mut cases = vec![];
+        let mut case_names = HashSet::new();
+        arbitrary_loop(u, 0, 100, |u| {
+            cases.push(self.arbitrary_named_type(u, &mut case_names)?);
+            Ok(true)
+        })?;
+
+        let default = if !cases.is_empty() && u.arbitrary()? {
+            let max_cases = u32::try_from(cases.len() - 1).unwrap();
+            Some(u.int_in_range(0..=max_cases)?)
+        } else {
+            None
+        };
+
+        Ok(VariantType { cases, default })
+    }
+
+    fn arbitrary_list_type(&mut self, u: &mut Unstructured<'a>) -> Result<ListType> {
+        Ok(ListType {
+            elem_ty: self.arbitrary_value_type(u)?,
+        })
+    }
+
+    fn arbitrary_tuple_type(&mut self, u: &mut Unstructured<'a>) -> Result<TupleType> {
+        let mut fields = vec![];
+        arbitrary_loop(u, 0, 100, |u| {
+            fields.push(self.arbitrary_value_type(u)?);
+            Ok(true)
+        })?;
+        Ok(TupleType { fields })
+    }
+
+    fn arbitrary_flags_type(&mut self, u: &mut Unstructured<'a>) -> Result<FlagsType> {
+        let mut fields = vec![];
+        let mut field_names = HashSet::new();
+        arbitrary_loop(u, 0, 100, |u| {
+            fields.push(crate::unique_string(100, &mut field_names, u)?);
+            Ok(true)
+        })?;
+        Ok(FlagsType { fields })
+    }
+
+    fn arbitrary_enum_type(&mut self, u: &mut Unstructured<'a>) -> Result<EnumType> {
+        let mut variants = vec![];
+        let mut variant_names = HashSet::new();
+        arbitrary_loop(u, 0, 100, |u| {
+            variants.push(crate::unique_string(100, &mut variant_names, u)?);
+            Ok(true)
+        })?;
+        Ok(EnumType { variants })
+    }
+
+    fn arbitrary_union_type(&mut self, u: &mut Unstructured<'a>) -> Result<UnionType> {
+        let mut variants = vec![];
+        let mut variant_names = HashSet::new();
+        arbitrary_loop(u, 0, 100, |u| {
+            variants.push(crate::unique_string(100, &mut variant_names, u)?);
```

Comment:
I may again be missing things from the spec, but I thought unions were unions of types rather than unions of strings? (basically the union/enum cases may be duplicated here by accident)

Reply:
They are, this is a bad copy paste. Good eye!